hack/generator/pkg/codegen/pipeline_remove_embedded_resources.go

```diff
@@ -0,0 +1,681 @@
+	renames := make(map[astmodel.TypeName]astmodel.TypeName)
+	for associated := range associatedNames {
+		embeddedName, err := parseContextualTypeName(associated)
+		if err != nil {
+			return nil, err
+		}
+		embeddedName.context = ""
+		renames[associated] = embeddedName.ToSimplifiedTypeName()
+	}
+
+	return renames, nil
+}
+
+func (r renamer) simplifyRemainingAssociatedNames(_ astmodel.TypeName, associatedNames astmodel.TypeNameSet) (map[astmodel.TypeName]astmodel.TypeName, error) {
+	// remove _0, which especially for the cases where there's only a single
+	// kind of usage will make the type name much clearer
+	renames := make(map[astmodel.TypeName]astmodel.TypeName)
+	for associated := range associatedNames {
+		embeddedName, err := parseContextualTypeName(associated)
+		if err != nil {
+			return nil, err
+		}
+
+		possibleRename := embeddedName.ToSimplifiedTypeName()
+		if !possibleRename.Equals(associated) {
+			renames[associated] = possibleRename
+		}
+	}
+
+	return renames, nil
+}
+
+func (r renamer) performRenames(
+	renames map[astmodel.TypeName]astmodel.TypeName,
+	flag astmodel.TypeFlag) (astmodel.Types, error) {
+
+	result := make(astmodel.Types)
+
+	renamingVisitor := astmodel.MakeTypeVisitor()
+	renamingVisitor.VisitTypeName = func(this *astmodel.TypeVisitor, it astmodel.TypeName, ctx interface{}) (astmodel.Type, error) {
+		if newName, ok := renames[it]; ok {
+			return astmodel.IdentityVisitOfTypeName(this, newName, ctx)
+		}
+		return astmodel.IdentityVisitOfTypeName(this, it, ctx)
+	}
+
+	for _, def := range r.types {
+		updatedDef, err := renamingVisitor.VisitDefinition(def, nil)
+		if err != nil {
+			return nil, err
+		}
+		// TODO: If we don't remove this, something is causing these types to not be emitted.
+		// TODO: Unsure what that is... should track it down
+		updatedType, err := flag.RemoveFrom(updatedDef.Type())
+		if err != nil {
+			return nil, err
+		}
+		result.Add(updatedDef.WithType(updatedType))
+	}
+
+	return result, nil
+}
+
+// simplifyTypeNames simplifies contextual type names if possible.
+func simplifyTypeNames(types astmodel.Types, flag astmodel.TypeFlag) (astmodel.Types, error) {
+	renames := make(map[astmodel.TypeName]astmodel.TypeName)
+
+	// Find all of the type names that have the flag we're interested in
+	updatedNames := make(map[astmodel.TypeName]astmodel.TypeNameSet)
+	for _, def := range types {
+		if flag.IsOn(def.Type()) {
+			embeddedName, err := parseContextualTypeName(def.Name())
+			if err != nil {
+				return nil, err
+			}
+
+			associatedNames := updatedNames[embeddedName.original]
+			associatedNames = associatedNames.Add(def.Name())
+			updatedNames[embeddedName.original] = associatedNames
+		}
+	}
+
+	renamer := renamer{types: types}
+	renameActions := []func(_ astmodel.TypeName, associatedNames astmodel.TypeNameSet) (map[astmodel.TypeName]astmodel.TypeName, error){
```

Comment:
Three suggestions.

Give the type `map[astmodel.TypeName]astmodel.TypeName` a meaningful name since we use it repeatedly. Perhaps "TypeAssociation`? It maybe should live in the `astmodel` package.

Give the type `func(astmodel.TypeName, astmodel.TypeNameSet) (map[astmodel.TypeName]astmodel.TypeName, error)` a meaningful name to make the array clearer. 

Give the methods more declarative names to indicate what they achieve - though I'm not sure what to suggest as I'm still working on understanding.

Reply:
Did all 3 of the suggested things above.